src/HydroShocktubeCMA/test_hydro_shocktube_cma.cpp

```diff
@@ -0,0 +1,236 @@
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 2) = 1 - 0.8 - 0.3 * pow(sin(20 * 3.14 * 0), 2);
+
+	} else if (i >= hi[0]) {
+		// x1 right-side boundary -- constant
+		for (int n = 0; n < numcomp; ++n) {
+			consVar(i, j, k, n) = 0;
+		}
+
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::gasEnergy_index) = P_R / (gamma - 1.);
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::gasInternalEnergy_index) = P_R / (gamma - 1.);
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::gasDensity_index) = rho_R;
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::x1GasMomentum_index) = 0.;
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::x2GasMomentum_index) = 0.;
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::x3GasMomentum_index) = 0.;
+
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 0) = 0.1;
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 1) = 0.3 * pow(sin(20 * 3.14 * 1), 2);
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 2) = 1 - 0.1 - 0.3 * pow(sin(20 * 3.14 * 1), 2);
+	}
+}
+
+template <> void RadhydroSimulation<ShocktubeProblem>::ErrorEst(int lev, amrex::TagBoxArray &tags, Real /*time*/, int /*ngrow*/)
+{
+	// tag cells for refinement
+
+	const Real eta_threshold = 0.1; // gradient refinement threshold
+	const Real rho_min = 0.01;	// minimum rho for refinement
+	auto const &dx = geom[lev].CellSizeArray();
+
+	for (amrex::MFIter mfi(state_new_cc_[lev]); mfi.isValid(); ++mfi) {
+		const amrex::Box &box = mfi.validbox();
+		const auto state = state_new_cc_[lev].const_array(mfi);
+		const auto tag = tags.array(mfi);
+
+		amrex::ParallelFor(box, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+			int const n = 0;
+			Real const rho = state(i, j, k, n);
+			Real const del_x = (state(i + 1, j, k, n) - state(i - 1, j, k, n)) / (2.0 * dx[0]);
+			Real const gradient_indicator = std::sqrt(del_x * del_x) / rho;
+
+			if (gradient_indicator > eta_threshold && rho >= rho_min) {
+				tag(i, j, k) = amrex::TagBox::SET;
+			}
+		});
+	}
+}
+
+auto problem_main() -> int
+{
+	// Problem parameters
+	const double max_time = 0.4;
+	const int max_timesteps = 8000;
+
+	// Problem initialization
+	const int ncomp_cc = Physics_Indices<ShocktubeProblem>::nvarTotal_cc;
+	amrex::Vector<amrex::BCRec> BCs_cc(ncomp_cc);
+	for (int n = 0; n < ncomp_cc; ++n) {
+		BCs_cc[0].setLo(0, amrex::BCType::ext_dir); // Dirichlet
+		BCs_cc[0].setHi(0, amrex::BCType::ext_dir);
+		for (int i = 1; i < AMREX_SPACEDIM; ++i) {
+			BCs_cc[n].setLo(i, amrex::BCType::int_dir); // periodic
+			BCs_cc[n].setHi(i, amrex::BCType::int_dir);
+		}
+	}
+
+	RadhydroSimulation<ShocktubeProblem> sim(BCs_cc);
+
+	// sim.cflNumber_ = CFL_number;
+	// sim.maxDt_ = max_dt;
+	// sim.initDt_ = initial_dt;
+	sim.stopTime_ = max_time;
+	sim.maxTimesteps_ = max_timesteps;
+
+	// Main time loop
+	sim.setInitialConditions();
+	sim.evolve();
+
+	// Compute test success condition
```

Comment:
You should add the conservation check here. There are other test problems that do something like this that you could use as an example.

Reply:
Do you remember which test implements such a check? `HydroShocktube` doesn't.

Reply:
`HydroBlast3D` checks energy conservation at the end of the run.

Reply:
Thanks. I implemented `computeAfterEvolve` following the energy conservation test in `HydroBlast3D`, and now I can see that CMA seems to be working very well. Error in mass conservation at the end of the run remains at machine precision! 

However, this only checks conservation at the end of the run. I would like to calculate and store the error after each timestep, so I can follow its time evolution (as shown in figure 3 of Plewa and Muller 1999). How do I do that?

Reply:
Here's an example of how to do that: https://github.com/quokka-astro/quokka/blob/de6c07b61aa54b0b5ef76a70502fffb4533eb144/src/RadMatterCoupling/test_radiation_matter_coupling.cpp#L111

Reply:
I implemented it, but I get an error:

```
/g/data1b/jh2/ps3459/quokka/src/HydroShocktubeCMA/test_hydro_shocktube_cma.cpp:228:26: error: ‘values’ was not declared in this scope
  228 |             specieSum += values.at(HydroSystem<ShocktubeProblem>::scalar0_index + n)[0];
```

`values` should come from the line: `auto [position, values] = fextract(state_new_cc_[0], Geom(0), 0, 0.5);` as in the Radiation matter coupling test right?

Reply:
Also, not sure if the line `auto [position, values] = fextract(state_new_cc_[0], Geom(0), 0, 0.5);` needs to be modified for the CMA test?

Reply:
No, that shouldn't need to be modified.

Reply:
Yes, that's where `values` should come from. I can't figure out this error without the larger context. Can you push your implementation of this?